pkg/api/core/v1/client/portforwarder.go

```diff
@@ -0,0 +1,304 @@
+// Copyright © 2021 - 2023 SUSE LLC
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/url"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/gorilla/websocket"
+)
+
+type ListenAddress struct {
+	Address     string
+	Protocol    string
+	FailureMode string
+}
+
+type ForwardedPort struct {
+	Local  uint16
+	Remote uint16
+}
+
+type ServicePortForwarder struct {
+	Client             *Client
+	Endpoint           string
+	ListeningAddresses []ListenAddress
+	Ports              []ForwardedPort
+	StopChan           <-chan struct{}
+}
+
+func NewServicePortForwarder(client *Client, endpoint string, addresses []string, ports []string, stopChan <-chan struct{}) (*ServicePortForwarder, error) {
+	if len(addresses) == 0 {
+		return nil, errors.New("you must specify at least 1 address")
+	}
+	parsedAddresses, err := parseAddresses(addresses)
+	if err != nil {
+		return nil, err
+	}
+	if len(ports) == 0 {
+		return nil, errors.New("you must specify at least 1 port")
+	}
+	parsedPorts, err := parsePorts(ports)
+	if err != nil {
+		return nil, err
+	}
+	return &ServicePortForwarder{
+		Client:             client,
+		Endpoint:           endpoint,
+		ListeningAddresses: parsedAddresses,
+		Ports:              parsedPorts,
+		StopChan:           stopChan,
+	}, nil
+}
+
+func parseAddresses(addressesToParse []string) ([]ListenAddress, error) {
```

Comment:
I would export this function, and put it at the very bottom, or in a different file, together with the `parsePorts`.

They are pretty "long" and can be easily unit tested.
Once we have some tests I think we will be able to simplify a bit the code with more confidence.

Reply:
I have exported these functions to the proxy.go file